### PR TITLE
[improvement](disk balance) Impr disk rebalancer sched when be load r…

### PR DESCRIPTION
**Pull request by deardeng**
…ebalancer idle

## Proposed changes

When use be load rebalancer, even after balanced, some backends load scores may be low or high. It will stop disk rebalancer working. Impr: when use be load rebalancer, if its cache empty for 10 min, then all the backends will try to do disk balance.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by airborne12 — APPROVED** [review]
LGTM

**Comment by dataroaring** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/clone/DiskRebalancer.java [hunk 1]
```diff
@@ -128,14 +129,17 @@ protected List<TabletSchedCtx> selectAlternativeTabletsForCluster(
...
-                    .getTabletScheduler().getRebalancer();
-            if (rebalancer != null && rebalancer.checkCacheEmptyForLong()) {
-                midBEs.addAll(lowBEs);
-                midBEs.addAll(highBEs);
-            }
-        } else if (!(lowBEs.isEmpty() && highBEs.isEmpty())) {
+        Rebalancer rebalancer = Env.getCurrentEnv().getTabletScheduler().getRebalancer();
+        List<Long> fromBes = new ArrayList<>();
+        List<Long> toBes = new ArrayList<>();
+        if (rebalancer != null && rebalancer.isNeedBalanced(clusterStat, medium, fromBes, toBes)) {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/DiskRebalancer.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java [hunk 2]
```diff
@@ -233,6 +233,21 @@ private boolean checkMoveCompleted(TabletMove move) {
...
+        fromBes.addAll(movesInProgress.get().asMap().values().stream()
+                .filter(p -> p.second != -1L).map(p -> p.first.fromBe).collect(Collectors.toList()));
+        toBes.addAll(movesInProgress.get().asMap().values().stream()
+                .filter(p -> p.second != -1L).map(p -> p.first.toBe).collect(Collectors.toList()));
+
+        return checkCacheEmptyForLong();
+    }
+
     // cache empty for 10 min
     public boolean checkCacheEmptyForLong() {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java [hunk 3]
```diff
@@ -77,6 +86,31 @@ public List<TabletSchedCtx> selectAlternativeTablets() {
     protected abstract List<TabletSchedCtx> selectAlternativeTabletsForCluster(
             LoadStatisticForTag clusterStat, TStorageMedium medium);
 
+    protected boolean isNeedBalanced(LoadStatisticForTag clusterStat, TStorageMedium medium,
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java [hunk 4]
```diff
@@ -55,6 +57,9 @@ public abstract class Rebalancer {
     // be id -> end time of prio
     protected Map<Long, Long> prioBackends = Maps.newConcurrentMap();
 
+    // tag -> (medium, timestamp)
+    private Table<LoadStatisticForTag, TStorageMedium, Long> lastPickTimeTable = HashBasedTable.create();
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java [hunk 5]
```diff
@@ -77,6 +86,31 @@ public List<TabletSchedCtx> selectAlternativeTablets() {
...
+    protected boolean isNeedBalanced(LoadStatisticForTag clusterStat, TStorageMedium medium,
+                                     List<Long> fromBes, List<Long> toBes) {
+        return true;
+    }
+
+    protected boolean unPickOverLongTime(LoadStatisticForTag clusterStat, TStorageMedium medium) {
+        if (lastPickTimeTable.containsRow(clusterStat)) {
+            Map<TStorageMedium, Long> mediumLastPickTimeMap = lastPickTimeTable.row(clusterStat);
+            if (mediumLastPickTimeMap.containsKey(medium)) {
+                Long lastPickTimeTable = mediumLastPickTimeMap.get(medium);
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 6]
```diff
@@ -1472,6 +1475,7 @@ private void finalizeTabletCtx(TabletSchedCtx tabletCtx, TabletSchedCtx.State st
         if (state == TabletSchedCtx.State.FINISHED) {
             tryAddAfterFinished(tabletCtx);
         }
+        rebalancer.removeFromLastPickTimeTable(tabletCtx.getTag(), tabletCtx.getStorageMedium());
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java [hunk 7]
```diff
@@ -77,6 +86,31 @@ public List<TabletSchedCtx> selectAlternativeTablets() {
...
+                return lastPickTimeTable > 0 && System.currentTimeMillis() > lastPickTimeTable + 10 * 60 * 1000L;
+            } else {
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+
+    protected void removeFromLastPickTimeTable(Tag tag, TStorageMedium medium) {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java]: fixed

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/clone/LoadStatisticForTag.java [hunk 8]
```diff
@@ -389,6 +389,16 @@ public void getBackendStatisticByClass(
 
         for (BackendLoadStatistic beStat : beLoadStatistics) {
             Classification clazz = beStat.getClazz(medium);
+            long debugHighBeId = DebugPointUtil.getDebugParamOrDefault("FE.HIGH_LOAD_BE_ID", -1L);
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/LoadStatisticForTag.java:392]: no need 392 ~ 401  ？LoadStatisticForTag.init has set it.

File: fe/fe-core/src/test/java/org/apache/doris/utframe/MockedBackendFactory.java [hunk 9]
```diff
@@ -248,6 +263,62 @@ private void handleCloneTablet(TAgentTaskRequest request, TFinishTaskRequest fin
                     finishTaskRequest.setFinishTabletInfos(tabletInfos);
                 }
 
+                private void handleStorageMediumMigrate(TAgentTaskRequest request, TFinishTaskRequest finishTaskRequest) {
```
Comment (yujun777) [inline fe/fe-core/src/test/java/org/apache/doris/utframe/MockedBackendFactory.java:266]: just update src and dest disk's data used and avaialable capactity is ok,  like handleCloneTablet

**Comment by dataroaring** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/clone/LoadStatisticForTag.java (hunk 8 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/LoadStatisticForTag.java]: removed

File: fe/fe-core/src/test/java/org/apache/doris/utframe/MockedBackendFactory.java (hunk 9 above)
Comment (deardeng) [reply fe/fe-core/src/test/java/org/apache/doris/utframe/MockedBackendFactory.java:258]: fixed

File: fe/fe-core/src/test/java/org/apache/doris/utframe/MockedBackendFactory.java [hunk 10]
```diff
@@ -248,6 +255,28 @@ private void handleCloneTablet(TAgentTaskRequest request, TFinishTaskRequest fin
...
+                    long dataSize = Math.max(1, CatalogTestUtil.getTabletDataSize(req.tablet_id));
+                    DiskInfo diskInfo = getDisk(req.data_dir);
+                    if (diskInfo != null) {
+                        diskInfo.setDataUsedCapacityB(Math.min(diskInfo.getTotalCapacityB(),
+                                diskInfo.getDataUsedCapacityB() + dataSize));
+                        diskInfo.setAvailableCapacityB(Math.max(0L,
+                                diskInfo.getAvailableCapacityB() - dataSize));
+                        diskInfo.setState(DiskState.ONLINE);
+                    }
+                    Map<String, DiskInfo> newDiskInfos = Maps.newHashMap();
```
Comment (yujun777) [inline fe/fe-core/src/test/java/org/apache/doris/utframe/MockedBackendFactory.java:270]: delete line 270 ~ 273;

also need modify the src disk' available and used

**Review by yujun777 — CHANGES_REQUESTED** [review]

**Review by yujun777 — APPROVED** [review]
LGTM

**Comment by deardeng** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**